pkg/sync/syncWorker.go

```diff
@@ -0,0 +1,205 @@
+	conn, err := grpc.NewClient(
+		addr,
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+		grpc.WithDefaultCallOptions(),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to connect to peer: %v", err)
+	}
+	log.Info(fmt.Sprintf("Successfully connected to peer at %s", addr))
+	return conn, nil
+}
+
+func (s *syncWorker) setupStream(node registry.Node, conn *grpc.ClientConn) (message_api.ReplicationApi_BatchSubscribeEnvelopesClient, error) {
+	client := message_api.NewReplicationApiClient(conn)
+	stream, err := client.BatchSubscribeEnvelopes(
+		s.ctx,
+		&message_api.BatchSubscribeEnvelopesRequest{
+			Requests: []*message_api.BatchSubscribeEnvelopesRequest_SubscribeEnvelopesRequest{
+				{
+					Query: &message_api.EnvelopesQuery{
+						Filter: &message_api.EnvelopesQuery_OriginatorNodeId{
+							OriginatorNodeId: node.NodeID,
+						},
+						LastSeen: nil,
+					},
+				},
+			},
+		},
+	)
+	if err != nil {
+		return nil, fmt.Errorf(
+			"Failed to batch subscribe to peer: %v",
+			err,
+		)
+	}
+	return stream, nil
+}
+
+func (s *syncWorker) listenToStream(stream message_api.ReplicationApi_BatchSubscribeEnvelopesClient) error {
+	for {
+		envs, err := stream.Recv()
+		// TODO(rich) Handle normal stream closure properly
+		if err != nil {
+			return fmt.Errorf(
+				"Stream closed with error: %v",
+				err)
+		}
+		for _, env := range envs.Envelopes {
+			s.insertEnvelope(env)
+		}
+	}
+}
+
+func (s *syncWorker) insertEnvelope(env *message_api.OriginatorEnvelope) {
+	log.Info(fmt.Sprintf("Replication server received envelope %s", env))
+	// TODO(rich) Validation logic - share code with API service and publish worker
+	originatorBytes, err := proto.Marshal(env)
+	if err != nil {
+		log.Error("Failed to marshal originator envelope", zap.Error(err))
+	}
+
+	unsignedEnvelope := &message_api.UnsignedOriginatorEnvelope{}
+	err = proto.Unmarshal(env.GetUnsignedOriginatorEnvelope(), unsignedEnvelope)
+	if err != nil {
+		log.Error(
+			"Failed to unmarshal unsigned originator envelope",
+			zap.Error(err),
+		)
+	}
+
+	clientEnvelope := &message_api.ClientEnvelope{}
+	err = proto.Unmarshal(
+		unsignedEnvelope.GetPayerEnvelope().GetUnsignedClientEnvelope(),
+		clientEnvelope,
+	)
+	if err != nil {
+		log.Error(
+			"Failed to unmarshal client envelope",
+			zap.Error(err),
+		)
+	}
+
+	q := queries.New(s.store)
+
+	inserted, err := q.InsertGatewayEnvelope(
+		s.ctx,
+		queries.InsertGatewayEnvelopeParams{
+			OriginatorNodeID: int32(unsignedEnvelope.GetOriginatorNodeId()),
+			OriginatorSequenceID: int64(
+				unsignedEnvelope.GetOriginatorSequenceId(),
+			),
+			Topic:              clientEnvelope.GetAad().GetTargetTopic(),
+			OriginatorEnvelope: originatorBytes,
+		},
+	)
+	if err != nil {
+		log.Error("Failed to insert gateway envelope", zap.Error(err))
+	} else if inserted == 0 {
+		// Envelope was already inserted by another worker
+		log.Error("Envelope already inserted")
```

Comment:
can we log this as `Debug`? Given that we have 2+ workers, it will be a very common scenario.

Reply:
How about I change it to `Warn`? The reason for splitting off the `syncService` and `syncWorker` is so that we can have a single worker per cluster, so it's unexpected to see this in normal operation

Reply:
We should probably apply the same logic to sync workers as we do to the grpc servers. Either we guarantee that there is only ever 0-1 (by shutting it down first and breaking sync for a few minutes) or we have to assume there are 2+.

Reply:
I think it makes sense to assume that there will be 2 during upgrades - that will log a few warnings here, and then nothing else in normal operation